developer/src/kmc-convert/src/keylayout-to-kmn/keylayout-to-kmn-converter.ts

```diff
@@ -1,27 +1,990 @@
+import { KeylayoutFileReader } from './keylayout-file-reader.js';
+
+export interface convert_object {
+  keylayout_filename: string,
+  kmn_filename: string,
+  arrayOf_Modifiers: string[][],
+  arrayOf_Rules: Rule[],
+};
 
 export class KeylayoutToKmnConverter {
+
   static readonly INPUT_FILE_EXTENSION = '.keylayout';
   static readonly OUTPUT_FILE_EXTENSION = '.kmn';
+  static readonly USED_KEYS_COUNT = 51;               // we use key Nr 0 (A) -> key Nr 49 (Space)
+  static readonly MAX_CTRL_CHARACTER = 32;
+  static readonly SKIP_COMMENTED_LINES = false;
+  static readonly KMC_CONVERT_VERSION = "0.1";
+
+  static readonly DATA_SUBFOLDER = 'data';
+  static readonly TEST_SUBFOLDER = 'test';
+
+  private options: CompilerOptions;
 
-  constructor(/*private*/ _callbacks: CompilerCallbacks, /*private*/ _options: CompilerOptions) {
-    // TODO: if these are needed, uncomment /*private*/ and remove _, and they will then
-    // be available as class properties
+  async init(callbacks: CompilerCallbacks, options: CompilerOptions): Promise<boolean> {
+    this.options = { ...options };
+    this.callbacks = callbacks;
+    return true;
   }
 
-  async run(inputFilename: string, outputFilename: string, binaryData: Uint8Array): Promise<ConverterToKmnArtifacts> {
-    if(!inputFilename || !outputFilename || !binaryData) {
+  constructor(private callbacks: CompilerCallbacks, options: CompilerOptions) { };
+
+  /**
+   * @brief  member function to run read/convert/write
+   * @param  inputFilename the ukelele .keylayout-file to be converted
+   * @param  outputFilename the resulting keyman .kmn-file
+   * @return null on success
+   */
+  async run(inputFilename: string, outputFilename: string = ""): Promise<ConverterToKmnArtifacts> {
+
+    if (!inputFilename) {
       throw new Error('Invalid parameters');
     }
 
-    console.error('TODO: implement KeylayoutToKmnConverter');
+    const KeylayoutReader = new KeylayoutFileReader(this.callbacks/*, this.options*/);
+    const jsonO: object = KeylayoutReader.read(inputFilename);
+    if (!jsonO) {
+      throw new Error('Error while processing read()');
+    }
+
+    const outArray: convert_object = await this.convert(jsonO, outputFilename);
+    if (!outArray) {
+      throw new Error('Error while processing convert()');
+    }
+
+    const kmnFileWriter = new KmnFileWriter(this.callbacks, this.options);
+    const out_text_ok: boolean = kmnFileWriter.write(outArray);
 
+    if (!out_text_ok) {
+      throw new Error('Error while processing write()');
+    }
     return null;
   }
-}
\ No newline at end of file
+
+  /**
+   * @brief  member function to read filename and behaviour of a json object into a convert_object
+   * @param  jsonObj containing filename, behaviour and rules of a json object
+   * @return an convert_object containing all data ready to print out
+   */
+  public convert(jsonObj: any, outputfilename: string = ""): convert_object {
```

Comment:
should be private; if you need access for tests, see the pattern at https://github.com/keymanapp/keyman/blob/9c1c41745704d55f1bcc67685261d2ca3f61504e/developer/src/kmc-keyboard-info/src/keyboard-info-compiler.ts#L672-L675